ENH: Add derivatives for BarycentricInterpolator

jcwhitehead opened the pull request:
#### What does this implement/fix?

Implements derivatives for `BarycentricInterpolator`.

I've extended the existing tests for the `KroghInterpolator` derivatives to also test the new methods, and added a few additional tests.

#### Reference issue

This should close https://github.com/scipy/scipy/issues/2725.

#### Additional information

I haven't contributed to scipy before – let me know if I've missed anything! Happy to implement changes before it can be merged.

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -541,33 +563,74 @@ class BarycentricInterpolator(_Interpolator1D):
+    >>> axs[0].legend(axs[0].get_lines(), labels, bbox_to_anchor=(0., 1.02, 1., .102),
+    ...               loc='lower left', ncols=3, mode="expand", borderaxespad=0., frameon=False)
+    >>> plt.tight_layout()
+    >>> plt.show()
```
![image](https://user-images.githubusercontent.com/60778417/227737478-e55a81ac-46b5-4e96-a9d6-726c2978d572.png)
Can you make the nodes easier to see against the lines. Also minor nit but the text in the legend overlaps

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -541,33 +563,74 @@ class BarycentricInterpolator(_Interpolator1D):
 
-    def __init__(self, xi, yi=None, axis=0):
-        _Interpolator1D.__init__(self, xi, yi, axis)
+    def __init__(self, xi, yi=None, axis=0, wi=None):
```
`wi` needs documenting

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -541,33 +563,74 @@ class BarycentricInterpolator(_Interpolator1D):
+                dist[inv_permute[i]] = 1.0
+                prod = np.prod(dist)
+                if prod == 0.0:
+                    raise ValueError("Interpolation points xi must be distinct.")
```
Can you add a test for this

jcwhitehead replied:
Just to check I understand - you mean a test which attempts to initialise a polynomial with some repeated interpolation points, and checks for the `ValueError`? If so, sure, I will do that.

j-bowhay replied:
Exactly, just to make sure this error does get raised when you expect it to

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +743,91 @@ def _evaluate(self, x):
+            Point or points at which to evaluate the derivatives
+        der : integer, optional
+            Which derivative to extract. This number includes the
+            function value as 0th derivative.
```
```suggestion
            function value as 0th derivative. Defaults to
            the first derivative.
```

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +743,91 @@ def _evaluate(self, x):
+        x : array_like
+            Point or points at which to evaluate the derivatives
+        der : integer, optional
+            Which derivative to extract. This number includes the
```
```suggestion
            Which derivative to evaluate. This number includes the
```

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +743,91 @@ def _evaluate(self, x):
+        Returns
+        -------
+        d : ndarray
+            Derivative interpolated at the x-points. Shape of d is
```
```suggestion
            Derivative interpolated at the x-points. Shape of ``d`` is
```

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -703,6 +852,11 @@ def barycentric_interpolate(xi, yi, x, axis=0):
         The y coordinates of the points the polynomial should pass through.
     x : scalar or array_like
         Points to evaluate the interpolator at.
+    der : int or list, optional
+        How many derivatives to extract: `None` for all potentially
```
```suggestion
    der : int or list or None, optional
        How many derivatives to extract: ``None`` for all potentially
```

tupui replied:
Usually we don't put backticks at all for None, but I am not sure we are very consistent here.

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -192,6 +196,35 @@ def test_high_derivative(self):
+        P = KroghInterpolator(self.xs, ys, axis=0)
+        D = P.derivatives(self.test_xs)
+        for i in range(D.shape[0]):
+            assert_almost_equal(np.stack((poly1.deriv(i)(self.test_xs),
```
New tests should use `np.assert_allclose`
```suggestion
            np.assert_allclose(np.stack((poly1.deriv(i)(self.test_xs),
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -192,6 +196,35 @@ def test_high_derivative(self):
+
+        P = KroghInterpolator(self.xs, ys, axis=0)
+        for i in range(P.n):
+            assert_almost_equal(np.stack((poly1.deriv(i)(self.test_xs),
```
```suggestion
            np.assert_allclose(np.stack((poly1.deriv(i)(self.test_xs),
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -309,6 +342,60 @@ def test_scalar(self):
+        P = BarycentricInterpolator(self.xs, self.ys)
+        D = P.derivatives(self.test_xs)
+        for i in range(D.shape[0]):
+            assert_almost_equal(self.true_poly.deriv(i)(self.test_xs), D[i])
```
```suggestion
            np.assert_allclose(self.true_poly.deriv(i)(self.test_xs), D[i])
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -309,6 +342,60 @@ def test_scalar(self):
+        P = BarycentricInterpolator(self.xs, self.ys)
+        D = P.derivatives(self.test_xs, len(self.xs)+2)
+        for i in range(D.shape[0]):
+            assert_almost_equal(self.true_poly.deriv(i)(self.test_xs), D[i])
```
```suggestion
            np.assert_allclose(self.true_poly.deriv(i)(self.test_xs), D[i])
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -309,6 +342,60 @@ def test_scalar(self):
+        m = 10
+        r = P.derivatives(self.test_xs, m)
+        for i in range(m):
+            assert_almost_equal(P.derivative(self.test_xs, i), r[i])
```
```suggestion
            np.assert_allclose(P.derivative(self.test_xs, i), r[i])
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -309,6 +342,60 @@ def test_scalar(self):
+    def test_high_derivative(self):
+        P = BarycentricInterpolator(self.xs, self.ys)
+        for i in range(len(self.xs), 2*len(self.xs)):
+            assert_almost_equal(P.derivative(self.test_xs, i),
```
```suggestion
            np.assert_allclose(P.derivative(self.test_xs, i),
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -309,6 +342,60 @@ def test_scalar(self):
+        P = BarycentricInterpolator(self.xs, ys, axis=0)
+        D = P.derivatives(self.test_xs)
+        for i in range(D.shape[0]):
+            assert_almost_equal(np.stack((poly1.deriv(i)(self.test_xs),
```
```suggestion
            np.assert_allclose(np.stack((poly1.deriv(i)(self.test_xs),
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -309,6 +342,60 @@ def test_scalar(self):
+
+        P = BarycentricInterpolator(self.xs, ys, axis=0)
+        for i in range(P.n):
+            assert_almost_equal(np.stack((poly1.deriv(i)(self.test_xs),
```
```suggestion
            np.assert_allclose(np.stack((poly1.deriv(i)(self.test_xs),
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -348,10 +443,21 @@ def test_shapes_1d_vectorvalue(self):
-        values = barycentric_interpolate(self.xs, self.ys, self.test_xs)
-        assert_almost_equal(P(self.test_xs), values)
+        bi = barycentric_interpolate
+        assert_almost_equal(P(self.test_xs), bi(self.xs, self.ys, self.test_xs))
+        assert_almost_equal(P.derivative(self.test_xs, 2),
+                            bi(self.xs, self.ys, self.test_xs, der=2))
+        assert_almost_equal(P.derivatives(self.test_xs, 2),
```
```suggestion
        np.assert_allclose(P(self.test_xs), bi(self.xs, self.ys, self.test_xs))
        np.assert_allclose(P.derivative(self.test_xs, 2),
                            bi(self.xs, self.ys, self.test_xs, der=2))
        np.assert_allclose(P.derivatives(self.test_xs, 2),
```

j-bowhay commented on scipy/interpolate/tests/test_polyint.py:
```diff
@@ -368,7 +474,7 @@ def test_large_chebyshev(self):
         # With just capacity scaling, n=1097 fails
         # With both capacity scaling and random permutation, n=30000 succeeds
-        n = 800
+        n = 1100
```
What is the reason for this change?

jcwhitehead replied:
As indicated in the pre-existing comment, for `n = 800` the test passes even if the random permutation is disabled, so at present it won't catch a regression.

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +743,91 @@ def _evaluate(self, x):
+        d : ndarray
+            Derivative interpolated at the x-points. Shape of d is
+            determined by replacing the interpolation axis in the
+            original array with the shape of x.
```
```suggestion
            original array with the shape of ``x``.
```

tupui replied:
We use single quotes for parameters of the given function. We only use double quotes for things like one liner code examples or some simple maths sometimes.

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +743,91 @@ def _evaluate(self, x):
+            return cn
+
+        if self._diff_cij is None:
+            # c[i,j] = xi[i] - xi[j]
```
Can you remove the commented-out code

jcwhitehead replied:
These comments aren't code - they use matrix index notation to indicate the state of the `c` array after each successive in-place manipulation (which otherwise requires some thought to decipher and compare to the Berrut & Trefethen equations).

j-bowhay reviewed (changes requested):
Thanks @jcwhitehead, didn't have time to check the maths in any detail but a few suggestions to start

tupui commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +742,91 @@ def _evaluate(self, x):
+        return self._diff_baryint._evaluate_derivatives(x, der-1, all_lower=False)
+
+
+def barycentric_interpolate(xi, yi, x, der=0, axis=0):
```
Quick note about backward compatibility, we must observe the current order so add `der` after `axis`. Also we prefer to add keywords only arguments.
```suggestion
def barycentric_interpolate(xi, yi, x, axis=0, *, der=0):
```

jcwhitehead replied:
Unfortunately this brings the signature of `barycentric_interpolate` into conflict with `krogh_interpolate(xi, yi, x, der=0, axis=0)`, which I was hoping to avoid - is there a way to resolve this?

jcwhitehead replied:
So if I've understood correctly, the idea would be to:
- add a deprecation warning to `krogh_interpolate` if called with positional arguments for `der` or `axis`;
- add a deprecation warning to `barycentric_interpolate` if called with a positional argument for `axis`;
- allow both to take keyword arguments according to the new signature `*, der=0, axis=0)`.

If so that seems very sensible and I can add the deprecation warnings to the PR.

tupui replied:
Please don't do anything in this PR. This is something we would first need to discuss and plan for. Not that simple I am afraid and either way we would want to have dedicated PRs for that 😅

But yes that might look like similar to what you outlined.

j-bowhay replied:
> Please don't do anything in this PR. This is something we would first need to discuss and plan for. Not that simple I am afraid and either way we would want to have dedicated PRs for that sweat_smile
> 
> But yes that might look like similar to what you outlined.

Agree let's stick with the suggested change for this PR otherwise it will hold up what is a nice enhancement. That process can be discussed/planned afterwards

tupui commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +742,91 @@ def _evaluate(self, x):
                 p[r[:-1]] = self.yi[r[-1]]
         return p
 
+    def derivative(self, x, der=1):
```
```suggestion
    def derivative(self, x, *, der=1):
```

jcwhitehead replied:
Here `derivative` is providing a concrete implementation of the (pre-existing) abstract method of the base class `_Interpolator1DWithDerivatives` which has signature `derivative(self, x, der=1)`, so shouldn't it match?

j-bowhay replied:
I would prefer to see this match the signature of the base class

tupui commented on scipy/interpolate/_polyint.py:
```diff
@@ -210,14 +210,34 @@ def derivative(self, x, der=1):
+            Array of shape ``(der, x.size, self.yi.shape[1])`` containing
+            the derivatives from 0 to der-1
+        """
+        raise NotImplementedError()
```
We should add a message and also test this.

jcwhitehead replied:
What should the test/message be?  Currently it matches the abstract method for `_evaluate()` in `_Interpolator1D` here:
```
def _evaluate(self, x):
        """
        Actually evaluate the value of the interpolator.
        """
        raise NotImplementedError()
```
but I can add a message to/test for both if that's something that needs testing.

jcwhitehead replied:
It looks like the same situation elsewhere in scipy is handled by making the abstract class a subclass of `ABC` and using an `@abstractmethod` decorator, together with:
```
raise NotImplementedError("This method is only implemented with an "
                          "associated child of the base class.")
```
which could be a reasonable message to re-use here.

j-bowhay replied:
This class is not part of the public API so I think we are fine to leave it as is

tupui commented on scipy/interpolate/_polyint.py:
```diff
@@ -210,14 +210,34 @@ def derivative(self, x, der=1):
+        ----------
+        x : array_like
+            1D array of points at which to evaluate the derivatives
+        der: integer, optional
```
```suggestion
        der : integer, optional
```

tupui reviewed (commented):
Thank you @jcwhitehead. Some additional suggestions to help with the review.

(Thanks @j-bowhay! A few things for you to note 😉)

jcwhitehead commented:
I just removed @ev-br as a requested reviewer by accident - apologies (I don't seem to have a way to undo this).

I think I've addressed all the points raised but if there's any more feedback I'm happy to iterate further. Please do just 'unresolve' a conversation I've marked as resolved if you disagree and think there's more to be done.

tupui commented:
> I just removed @ev-br as a requested reviewer by accident - apologies (I don't seem to have a way to undo this).

No worries, this has no impact for us and how we approach PR reviews. (We don't use these features that much.)

> I think I've addressed all the points raised but if there's any more feedback I can keep iterating further. Please do just 'unresolve' a conversation I've marked as resolved if you disagree and think there's more to be done.

In general please don't resolve discussions at all. This is important for maintainers to know where they are at in their review. You can add a quick comment or some thumbs up if you want on specific discussions to indicate you took care of it. Or make a general message listing what you did and what's left to be done.

tupui commented on scipy/interpolate/_polyint.py:
```diff
@@ -190,43 +190,63 @@ def derivatives(self, x, der=None):
+        x : array_like
+            1D array of points at which to evaluate the derivatives
+        der : integer, optional
+            The number of derivatives to evaluate, from 'order 0' (der=1) to order der-1.
```
Our linter is a bit lax at the moment, but we still would like to have everything under 80 chars (79 max).

jcwhitehead commented on scipy/interpolate/_polyint.py:
```diff
@@ -541,49 +570,91 @@ class BarycentricInterpolator(_Interpolator1D):
-    def __init__(self, xi, yi=None, axis=0):
-        _Interpolator1D.__init__(self, xi, yi, axis)
+    def __init__(self, xi, yi=None, axis=0, *, wi=None):
+        _Interpolator1DWithDerivatives.__init__(self, xi, yi, axis)
```
Should the corresponding line in `KroghInterpolator` also be updated to use `super()` or is this kind of style change only for new code?

jcwhitehead commented on scipy/interpolate/_polyint.py:
```diff
@@ -665,8 +738,8 @@ def _evaluate(self, x):
-            c = self.wi/c
-            with np.errstate(divide='ignore'):
+            c = self.wi / c
+            with np.errstate(divide='ignore', invalid='ignore'):
```
Thanks for spotting this - I think this is probably now unnecessary after I added the `ValueError` in `__init__` to catch the special (nonsensical) case in which nodes coincide, so I will remove it.

jcwhitehead commented:
Those latest three CI failures don't look like they're related to anything that's changed in the commit.

From my side I think everything raised so far has now been addressed - if not or if there's more feedback I'm happy to keep iterating to get it approved and merged.

j-bowhay commented:
Thanks @jcwhitehead this is on my list to have another look at but it might be a couple of weeks before I get enough time

j-bowhay reviewed (changes requested):
A few minor changes but otherwise I think this is looking good. See https://numpydoc.readthedocs.io/en/latest/format.html#common-rest-concepts for when to use single vs double ticks.

Finally, as this is a new feature please could you write a post to the mailing list inviting others to take a look.

jcwhitehead commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +753,94 @@ def _evaluate(self, x):
+        return self._finish_y(y, x_shape)
+
+    def _evaluate_derivatives(self, x, der=None, all_lower=True):
+        # NB: der here is not the order of the highest derivative;
```
This is just the meaning of `der` within the existing `_Interpolator1DWithDerivatives` class, not something I've chosen or added. The implementation here has to match for the class methods to work.

I agree it's misleading, which is why I added the comment in case any

It could easily be renamed as this function isn't in the public API, but I'm not sure what the procedure is for that kind of thing.

j-bowhay replied:
Ok if there is a precedent then it's fine to leave

jcwhitehead commented on scipy/interpolate/_polyint.py:
```diff
@@ -541,49 +572,92 @@ class BarycentricInterpolator(_Interpolator1D):
+    >>> P = BarycentricInterpolator(xi, f(xi))
+    >>> fig, axs = plt.subplots(5, 1, sharex=True, layout='constrained', figsize=(7,10))
+    >>> x = np.linspace(0, np.pi, 100)
+    >>> axs[0].plot(x, P(x), 'r:', x, f(x), 'k--', xi, f(xi), 'xk')
+    >>> axs[1].plot(x, P.derivative(x), 'r:', x, f_d1(x), 'k--', xi, f_d1(xi), 'xk')
+    >>> axs[2].plot(x, P.derivative(x, 2), 'r:', x, f_d2(x), 'k--', xi, f_d2(xi), 'xk')
+    >>> axs[3].plot(x, P.derivative(x, 3), 'r:', x, f_d3(x), 'k--', xi, f_d3(xi), 'xk')
+    >>> axs[4].plot(x, P.derivative(x, 4), 'r:', x, f_d4(x), 'k--', xi, f_d4(xi), 'xk')
```
<img src="https://user-images.githubusercontent.com/25435662/230965799-58eea24a-7e58-4fcd-82f9-f28094b19bfd.png" width="50%">

The nodes are very clear already - not sure what you mean here.

j-bowhay replied:
Fine, lets leave as is

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +753,94 @@ def _evaluate(self, x):
+        if self._diff_cij is None:
+            # c[i,j] = xi[i] - xi[j]
+            c = self.xi[..., np.newaxis] - self.xi
+            # avoid division by 0 (diagonal entries are so far zero by construction)
+            np.fill_diagonal(c, 1)
+            # c[i,j] = w[j] / (xi[i] - xi[j])
+            c = self.wi / c
+
+            # c[i,j] = (w[j] / w[i]) / (xi[i] - xi[j])
+            c = c / self.wi[..., np.newaxis]
+
+            # fill in correct diagonal entries: each column sums to 0
+            np.fill_diagonal(c, 0)
```
Here we can we use np.divide(..., where=c!=0) to avoid having to rewrite the diagonal twice

j-bowhay commented:
Friendly ping @jcwhitehead, are you interested in finishing this up? This is looking in good shape with only some minor changes required so it would be great to see this in the next SciPy release

j-bowhay commented:
@ev-br @Kai-Striega @AtsushiSakai I have pushed some changes to hopefully get this over the line. Since I have made some edits would one of you be able to have a look over this

ev-br commented on scipy/interpolate/_polyint.py:
```diff
@@ -683,8 +758,89 @@ def _evaluate(self, x):
+            c = self.xi[..., np.newaxis] - self.xi
+            # c[i,j] = (w[j] / w[i]) / (xi[i] - xi[j])
+            c = np.divide(self.wi, c * self.wi[..., np.newaxis],
+                          out=np.zeros_like(c), where=c != 0)
```
Not sure what is going on here. `out=np.zeros_like(c)` signals "throw away the result", but then it is returned, ouch. It is probably deliberately ingenious (which it is!), but can we please write it in a more boring way?

j-bowhay replied:
This was one of the changes I made to avoid rewriting the diagonal back a forth (as c is 0 along the diagonal). Here the division is performed normally apart from where `c==0` where the result is instead left unchanged in the `out` array (hence is zero as desired).

Would you be happy with a comment explaining what is happening?

jcwhitehead replied:
Unfortunately I haven't had any time to keep up with this lately so I'm out of the loop, but I'd actually be a bit surprised if the new version with its quadratic comparisons was faster than the previous version with two linear fills. 

If it's a meaningful difference it's probably worth propagating to the older barycentric interpolation code which uses the same two diagonal fills.

@j-bowhay I guess you've checked the new version is faster - do you have some numbers?

j-bowhay replied:
I have to admit I thought I benchmarked this but clearly not. There is no applicable difference so I will revert the change:
```
import numpy as np
import perfplot

def setup(n):
    xi = np.random.random(n)
    c =  xi - xi[:, np.newaxis]
    np.fill_diagonal(c, 0)
    return c

def old(x):
    c = np.copy(x)
    np.fill_diagonal(c, 1)
    c  = 1/c
    np.fill_diagonal(c,0)
    return c

def new(x):
    c = np.copy(x)
    return np.divide(1, c, out=np.zeros_like(c), where=c!=0)

perfplot.show(setup=setup, kernels=[old, new], n_range=[2**k for k in range(15)])
"""
![image](https://github.com/scipy/scipy/assets/60778417/518852e4-8a97-4751-8497-62dc4497b076)

ev-br replied:
I'm fine with just adding a comment, FTR.

ev-br commented on scipy/interpolate/_polyint.py:
```diff
@@ -683,8 +758,89 @@ def _evaluate(self, x):
+
+        if self._diff_cij is None:
+            # c[i,j] = xi[i] - xi[j]
+            c = self.xi[..., np.newaxis] - self.xi
```
Nit: The comment hints that `self.xi` is 1D, is this always correct? If it is, I'd ask to consider writing a more explicit `xi[:, np.newaxis]`. 

Rationale: an Ellipsis signals an unspecified number of multiple dimensions. So if it really possibly multidim, could you please make it more explicit in the comment line above.

j-bowhay replied:
Ah yes good spot `self.xi` is the x coordinates of the points the polynomial should pass through so is 1d

ev-br commented on scipy/interpolate/_polyint.py:
```diff
@@ -677,8 +753,94 @@ def _evaluate(self, x):
+            # fill in correct diagonal entries: each column sums to 0
+            np.fill_diagonal(c, 0)
+            # c[j,j] = -sum_{i != j} c[i,j]
+            d = -c.sum(axis=1)
```
Equation numbers of the original paper would be great to add actually.

ev-br reviewed (commented):
LGTM modulo minor nits.

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -547,49 +577,92 @@ class BarycentricInterpolator(_Interpolator1D):
+            # these numerical stability improvements will be able to provide all
+            # the points to the constructor.
+            self._inv_capacity = 4.0 / (np.max(self.xi) - np.min(self.xi))
+            permute = np.random.permutation(self.n)
```
The unseeded randomness predates this pr however happy to tackle it

ev-br replied:
Ah, ok to leave then.

j-bowhay commented:
Thanks @ev-br @jcwhitehead, I have hopefully addressed all comments including reverting the use of `np.divide` and adding a `random_state` argument

ev-br commented:
This now LGTM. Anything you want to add/tweak @jcwhitehead @j-bowhay ? 

One minor hiccup is 11 commits looking a bit too much here, would be nice to squash some, while preserving authorship.

j-bowhay commented on scipy/interpolate/_polyint.py:
```diff
@@ -190,43 +191,64 @@ def derivatives(self, x, der=None):
     def derivative(self, x, der=1):
         """
-        Evaluate one derivative of the polynomial at the point x
+        Evaluate a single derivative of the polynomial at the point x.
```
```suggestion
        Evaluate a single derivative of the polynomial at the point `x`.
```

j-bowhay commented:
I had one final look and all happy from my end.

I am happy just to squash merge if that works for everyone (my contributions were mostly just housekeeping)

ev-br commented:
OK, let's merge this then if everyone's happy. Thank you @jcwhitehead , @j-bowhay !
Fun fact: the original issue this PR closed was opened exactly one day short of ten years ago :-). Great to see it fixed!

P.S. Turns out GH adds a co-authored-by stanza to commit messages on squash-merges.

j-bowhay commented:
Thanks @jcwhitehead for getting the ball rolling on this, it was a great first contribution. I hope we can tempt you back for more when you have the time!
